jsonschema/validators.py

```diff
@@ -259,17 +267,40 @@ def from_schema(cls, schema, *args, **kwargs):
 
         return cls(schema.get(u"id", u""), schema, *args, **kwargs)
 
+    def push_scope(self, scope):
+        self.scopes_stack.append(
+            self.urljoin_cache(self.resolution_scope, scope))
+
+    def pop_scope(self):
```

Comment:
Since these (pop/push) are being called from `ref()` I think they need to remain public.  I'm still not sure about the composition route.

The errors here would be for `pop()` on an empty stack?  Theoretically that should never happen because we always match push/pop, but a nicer error won't hurt either.